File: ARC/arc175/a.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

const int MOD = 998244353;

// DFS関数
int dfs(const vector<int>& P, string& S, int index, vector<bool>& taken) {
    if (index == S.size()) {
        // 全員がスプーンを取っているかチェック
        for (bool t : taken) {
            if (!t) return 0;
        }
        return 1;
    }

    int count = 0;
    int i = P[index] - 1;
    if (S[i] != '?') {
        // 利き手が決まっている場合
        int spoon = (S[i] == 'L') ? i : (i + 1) % S.size();
        if (!taken[spoon]) {
            taken[spoon] = true;
            count = dfs(P, S, index + 1, taken);
            taken[spoon] = false;
        }
    } else {
        // 利き手が不明な場合
        for (char hand : {'L', 'R'}) {
            int spoon = (hand == 'L') ? i : (i + 1) % S.size();
            if (!taken[spoon]) {
                taken[spoon] = true;
                S[i] = hand;
                count = (count + dfs(P, S, index + 1, taken)) % MOD;
                S[i] = '?';
                taken[spoon] = false;
            }
        }
    }
    return count;
}

int main() {
    int N;
    cin >> N;
    vector<int> P(N);
    for (int& p : P) cin >> p;
    string S;
    cin >> S;

    vector<bool> taken(N, false);
    cout << dfs(P, S, 0, taken) << endl;
    return 0;
}

File: ARC/arc175/a.py

N = int(input())
P = [0] + list(map(int, input().split()))
S = input()

S = ' ' + S

import math

mod = 998244353

# 木構造を使用する

# もし，左利きなら左の子の値を自身の値にする
# もし，右利きなら右の子の値を自身の値にする
# どちらでもないなら，左右の子の値の和を自身の値にする
# 末端のノードは1を返す

cnt = 0

spoon = [None] + [1] * (N + 1) # スプーンiが残っているかどうか


def dfs(v):
    

    # もし，末端のノードなら，spoonが取れるかどうかを返す
    if v >= 2**N:
        
        idx = int(math.log2(v)) + 1
        
        if S[idx] == 'L':
            if spoon[idx] == 1:
                spoon[idx] = 0
                return 1
            else:
                return 0
            
        elif S[idx] == 'R':
            if spoon[(idx+1)%N] == 1:
                spoon[(idx+1)%N] = 0
                return 1
            else:
                return 0
        
        else:
            if spoon[idx] == 1:
                spoon[idx] = 0
                left = 1
                spoon[idx] = 1
                if spoon[(idx+1)%N] == 1:
                    spoon[(idx+1)%N] = 0
                    right = 1
                    spoon[(idx+1)%N] = 1
                    return left + right
                else:
                    return left
            else:
                left = 0
                if spoon[(idx+1)%N] == 1:
                    spoon[(idx+1)%N] = 0
                    right = 1
                    spoon[(idx+1)%N] = 1
                    return left + right
                else:
                    return left
                
    # そうでないなら，左右の子の値を計算して，自身の値を返す
    else :
        
        idx = int(math.log2(v)) + 1
        
        # 左利きの場合
        if S[idx] == 'L':
            spoon[idx] = 0
            tmp = dfs(v * 2)
            spoon[idx] = 1
            return tmp
        
        # 右利きの場合
        elif S[idx] == 'R':
            spoon[(idx+1)%N] = 0
            tmp = dfs(v * 2 + 1)
            spoon[(idx+1)%N] = 1
            return tmp
    
        # どちらでもない場合
        else:
            idx = int(math.log2(v)) + 1
            spoon[idx] = 0
            left = dfs(v * 2)
            spoon[idx] = 1
            
            spoon[(idx+1)%N] = 0
            right = dfs(v * 2 + 1)
            spoon[(idx+1)%N] = 1
            
            return (left + right) % mod
        
print(dfs(1))
    
    
    
    
    